rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,1173 @@
+{
+    fn enqueue(&self, mut packet: P) -> Result<(), Error> {
+        packet.set_timestamp(Instant::now());
+        self.ingress
+            .send(packet)
+            .map_err(|_| Error::ChannelError("Data channel is closed.".to_string()))?;
+        Ok(())
+    }
+}
+
+pub struct TokenBucketCellEgress<P>
+where
+    P: Packet,
+{
+    egress: mpsc::UnboundedReceiver<P>,
+    max_size: u32,
+    token_bucket: TokenBucket,
+    peak_token_bucket: Option<TokenBucket>,
+    packet_queue: DropTailQueue<P>,
+    next_available: Instant,
+    last_check: Instant,
+    config_rx: mpsc::UnboundedReceiver<TokenBucketCellConfig>,
+    timer: Timer,
+    state: AtomicI32,
+}
+
+impl<P> TokenBucketCellEgress<P>
+where
+    P: Packet + Send + Sync,
+{
+    fn set_config(&mut self, config: TokenBucketCellConfig) {
+        let now = Instant::now();
+        if let Some(token_bucket) = config.token_bucket {
+            info!(
+                "Previous next_available distance: {:?}",
+                self.next_available - now
+            );
+            info!("Previous check_point distance: {:?}", now - self.last_check);
+
+            // update tokens
+            let token_bucket = TokenBucket::from(token_bucket);
+            let length = time_to_length(
+                Duration::from_nanos(self.token_bucket.tokens as u64) + (now - self.last_check),
+                self.token_bucket,
+            );
+            let mut toks = length_to_time(length as usize, token_bucket).as_nanos() as i64;
+            let mut ptoks = 0;
+            if self.peak_token_bucket.is_some() {
+                let peak_token_bucket = self.peak_token_bucket.unwrap();
+                ptoks = peak_token_bucket.tokens + (now - self.last_check).as_nanos() as i64;
+                ptoks = if ptoks > peak_token_bucket.buffer * 1000000 {
+                    peak_token_bucket.buffer * 1000000
+                } else {
+                    ptoks
+                };
+            }
+
+            toks = if toks > token_bucket.buffer * 1000000 {
+                token_bucket.buffer * 1000000
+            } else {
+                toks
+            };
+
+            // modify self.token_bucket
+            self.last_check = now;
+            self.token_bucket.buffer = token_bucket.buffer;
+            self.token_bucket.rate = token_bucket.rate;
+            self.token_bucket.tokens = toks;
+
+            if self.peak_token_bucket.is_some() {
+                let peak_token_bucket = self.peak_token_bucket.unwrap();
+                self.peak_token_bucket = Some(TokenBucket {
+                    buffer: peak_token_bucket.buffer,
+                    rate: peak_token_bucket.rate,
+                    tokens: ptoks,
+                });
+            }
+
+            // calculate max_size
+            self.max_size = calculate_max_size(self.token_bucket, self.peak_token_bucket);
+
+            // check if packet can be delivered
+            self.next_available = Instant::now();
```

Comment:
Maybe calling this variable `next_available` is not so correct here. In `set_config` tokens of buckets will be **updated** according to time past since `last_check`. Also, if the rate of the bucket is modified, old `next_available` may become **inaccurate** because I calculated it according to old rate of the bucket when `last_check`. I just want to check if updated tokens can deliver a packet, and calculate and update `next_available` to a new accurate time. That's why I set `next_available` to the instant the configuration is modified, so that these checks and calculations will be done in `dequeue`. Since `dequeue` will not return anyting if tokens are not enough to send the first packet in the queue, I think there is no problem returning to `dequeue` even if it is not available for sending actually? Or is there a better way to implement these checks and caculations mentioned above?

Reply:
> Since dequeue will not return anyting if tokens are not enough to send the first packet in the queue, I think there is no problem returning to dequeue even if it is not available for sending actually

I shall correct my explanation. Actually, since there is a big `loop` on the top of the whole `enqueue`, it will not return until an explicit `return` sentence.

Reply:
I raise this point because `next_available` is intended to represent the next available time for sending. Even if this particular assignment isn't used and will be corrected by subsequent functions, it's crucial for each variable to align with its intended logical meaning when assigned. Failing to do so can complicate future code maintenance. If this assignment serves other purposes, such as triggering checks, it would be more appropriate to trigger such actions through other means rather than through this assignment. You can introduce more variables for such purposes.

Reply:
Also, why you use `Instant::now` instead of the `now` here? It is better to use the same timestamp in one function.